dap-launch.el

```diff
@@ -0,0 +1,241 @@
+that function is called with `funcall', and its result, which
+must be a string, is used in place of the variable. If you used
+capture groups in REGEX, the function you specified in VALUE is
+called with the variable as its only argument. This way, you can
+use `string-match' to get the capture groups. If, however, REGEX
+does not contain capture groups, your function is called without
+any arguments. Otherwise, if it is a symbol, the symbol's value
+is used the same way. Lastly, if it is a string, the string is
+used as a replacement. If no regex matches, the empty string is
+used as a replacement and a warning is issued.
+
+See `dap--launch-json-getenv' for an example on how to use
+capture groups in REGEX.")
+
+(defun dap--launch-json-eval-poly-type (value var)
+  "Get the value from VALUE depending on its type.
+If it is a function, and VAR is not nil, call VALUE and pass VAR as an argument.
+If it is a symbol, return its value.
+Otherwise, return VALUE"
+  (cond ((and var (functionp value)) (funcall value var))
+        ((functionp value) (funcall value))
+        ((symbolp value) (symbol-value value))
+        (t value)))
+
+(defun dap--warn-var-nil (var)
+  "Warn that VAR is an unknown launch.json variable."
+  (dap--launch-json-warn-nil
+   (format "variable '%s' is unknown and was ignored" var)))
+
+(defun dap-expand-variable (var)
+  "Expand variable VAR using `dap--launch-json-variables'."
+  (catch 'ret
+    (save-match-data
+      (dolist (var-pair dap-launch-json-variables)
+        (when (string-match (car var-pair) var)
+          (throw 'ret
+                 (or
+                  (dap--launch-json-eval-poly-type
+                   (cdr var-pair)
+                   (if (= (length (match-data)) 2) ;; no capture groups
+                       nil
+                     var
+                     (message "wow capture groups")))
+                  (dap--warn-var-nil var)
+                  "")))))
+    nil))
+
+(defun dap-expand-variables-in-string (s)
+  "Expand all launch.json variables of the from ${variable} in S.
+Return the result."
+  (let ((old-buffer (current-buffer)))
+    (with-temp-buffer
+      (insert s)
+      (goto-char (point-min))
+
+      (save-match-data
+        (while (re-search-forward "${\\([^}]*\\)}" nil t)
+          (let ((var (match-string 1)))
+            (replace-match
+             (with-current-buffer old-buffer
+               (dap-expand-variable var))))))
+
+      (buffer-string))))
+
+(defun dap--launch-json-expand-vars (conf)
+  "Non-destructively expand all variables in all strings of CONF.
+CONF is regular dap-mode launch configuration. Return the result."
+  (cond ((listp conf)
+         (apply #'nconc (cl-loop for (k v) on conf by #'cddr collect
+                                 (list k (dap--launch-json-expand-vars v)))))
+        ((stringp conf) (dap-expand-variables-in-string conf))
+        (t conf)))
+
+(defun dap--launch-json-prompt-configuration ()
```

Comment:
I agree. My approach is a bit ad-hoc. I propose that we add a plugin architecture to dap-debug. It should then call something like (dap-gather-templates) instead of using dap-debug-template-configurations in the interactive block. That function would then iterate over something like dap-backends-alist and gather all templates. I don't see any other references to that variable (aside from defcustoms) so that shouldn't be a problem?

Reply:
I am usually against introducing additional layers of abstraction when it is not clear what the benefit is but I guess in this case having something like won't affect the complexity of the code. 
```
(-mapcat #'funcall dap-templates-functions)
```

Reply:
Yes, that was my idea. I will implement this.

Reply:
Implemented in latest master of my fork.